### PR TITLE
create separate plugin for checkpointing

### PR DESCRIPTION
**Pull request by psychocoderHPC**
This pull request adds a plugin which controls the checkpoint creation and restart which was before a part of the hdf5 and adios plugin.
This is the first step to refactor our IO plugins to allow the usage as multi plugin.
It is now possible to restart from adios and create new checkpoints with hdf5.
**This plugin renames all command line options to control the checkpoint creation and restart.**

- add checkpoint plugin 
  - add interface to unify usage of IO-backends (adios or hdf5)
  - add Checkpoint plugin
  - register `Checkpoint` as plugin
-  refactor adios and hdf writer interfaces 
  - Refactor interfaces that both plugins can be used as stand allown plugin and controlled checkpoint plugin.
- **rename command line parameter**
  - `--checkpoints` to `--checkpoint.period`
  - `--softRestarts` to `--checkpoint.restart.soft`
  - `--restart` to `--checkpoint.restart`
  - `--restart-directory` to `--checkpoint.restart.directory`
  - `--restart-step` to `--checkpoint.restart.step`
  - `--checkpoint-directory` to `checkpoint.directory`

# Test

- [x] compile without Adios/HDF5
- [x] restart from hdf5 -> than create checkpoints with adios
- [x] restart from adios -> than create checkpoints HDF5

# Help message if Adios and HDF5 is available

```
  --checkpoint.restart                  Restart simulation
  --checkpoint.restart.directory arg (=checkpoints)
                                        Directory containing checkpoints for a 
                                        restart
  --checkpoint.restart.step arg         Checkpoint step to restart from
  --checkpoint.period arg               Period for checkpoint creation
  --checkpoint.directory arg (=checkpoints)
                                        Directory for checkpoints
...

Checkpoint:
  --checkpoint.backend arg              Optional backend for checkpointing 
                                        [adios, hdf5] default: adios
  --checkpoint.file arg                 Optional checkpoint filename (prefix)
  --checkpoint.restart.backend arg      Optional backend for restarting [adios,
                                        hdf5] default: adios
  --checkpoint.restart.file arg         checkpoint restart filename (prefix)
  --checkpoint.restart.chunkSize arg (=1000000)
                                        Number of particles processed in one 
                                        kernel call during restart to prevent 
                                        frame count blowup
  --checkpoint.adios.aggregators arg (=0)
                                        Number of aggregators [0 == number of 
                                        MPI processes]
  --checkpoint.adios.ost arg (=1)       Number of OST
  --checkpoint.adios.disable-meta       Disable online gather and write of a 
                                        global meta file, can be time consuming
                                        (use `bpmeta` post-mortem)
  --checkpoint.adios.transport-params arg
                                        additional transport parameters, see 
                                        ADIOS manual chapter 6.1.5, e.g., 
                                        'random_offset=1;stripe_count=4'
  --checkpoint.adios.compression arg (=none)
                                        ADIOS compression method, e.g., zlib 
                                        (see `adios_config -m` for help)
```

### CODE REVIEW

File: include/picongpu/plugins/output/IIOBackend.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+public:
+
+
+    IIOBackend()
+    {
+
+    }
+
+    template< typename T_IOBackend >
+    static std::shared_ptr<IIOBackend> create( bool const isIndependent = true )
```
Comment (ax3l) [inline include/picongpu/plugins/output/IIOBackend.hpp]: pls comment the `@param isIndependent paramter` :)

File: include/picongpu/plugins/output/IIOBackend.hpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#include "picongpu/plugins/ILightweightPlugin.hpp"
+
+#include <string>
+#include <memory>
+
+namespace picongpu
+{
+
+//! Interface for IO-backends with restart capability
+class IIOBackend : public ILightweightPlugin
```
Comment (ax3l) [inline include/picongpu/plugins/output/IIOBackend.hpp]: class needs to be intended by one level

File: include/picongpu/plugins/output/IIOBackend.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
+/* Copyright 2013-2017 Axel Huebl, Felix Schmitt, Heiko Burau, Rene Widera,
```
Comment (ax3l) [inline include/picongpu/plugins/output/IIOBackend.hpp]: only.... you ♬ ♪ :)

File: include/picongpu/plugins/output/IIOBackend.hpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    ) = 0;
+
+    virtual void expandHelp(
+        std::string const & prefix,
+        boost::program_options::options_description & desc
+    ) = 0;
+
+};
+
+} //namespace picongpu
```
Comment (ax3l) [inline include/picongpu/plugins/output/IIOBackend.hpp]: 1 space after `//`

File: docs/source/usage/plugins/checkpoint.rst [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                                   If relative, checkpoint files are stored under ``simOutput/<checkpoint-directory>``.
+                                   Default depends on the selected IO-backend.
+``--checkpoint.restart.backend``   IO-backend used to load a existent checkpoint.
+``--checkpoint.restart.file``      Relative or absolute fileset prefix for reading checkpoints.
+                                   If relative, checkpoint files are searched under ``simOutput/<checkpoint-directory>``.
+                                   Default depends on the selected IO-backend``.
+``--checkpoint.restart.chunkSize`` Number of particles processed in one kernel call during restart to prevent frame count blowup.
+``--checkpoint.<IO-backend>.*      Additional options to control the IO-bakend
+================================== ======================================================================================
+
```
[the remark this answers is not in the document]
Comment (psychocoderHPC) [reply docs/source/usage/plugins/checkpoint.rst]: added

File: docs/source/usage/plugins/checkpoint.rst [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+``--checkpoint.backend``           IO-backend used to create the checkpoint.
+``--checkpoint.file``              Relative or absolute fileset prefix for writing checkpoints.
+                                   If relative, checkpoint files are stored under ``simOutput/<checkpoint-directory>``.
+                                   Default depends on the selected IO-backend.
+``--checkpoint.restart.backend``   IO-backend used to load a existent checkpoint.
+``--checkpoint.restart.file``      Relative or absolute fileset prefix for reading checkpoints.
+                                   If relative, checkpoint files are searched under ``simOutput/<checkpoint-directory>``.
+                                   Default depends on the selected IO-backend``.
+``--checkpoint.restart.chunkSize`` Number of particles processed in one kernel call during restart to prevent frame count blowup.
+``--checkpoint.<IO-backend>.*      Additional options to control the IO-bakend
```
Comment (ax3l) [inline docs/source/usage/plugins/checkpoint.rst]: backend -.-

**Pull request merged**